### PR TITLE
Adding readWriteThis(ch: channel) in DistBag record

### LINKED ISSUE #15388: Add proc readWriteThis(ch: channel) for record DistBag in DistributedBag module.
The function to pass the elements stored in the Bag onto a channel.

```
use DistributedBag;

var bag = new DistBag(int);
bag.addBulk(1..5);
writeln(bag);
```
Current Output:
```(_impl = nil, _pid = 0, _rc = {_pid = 0})```

The above snippet should output ```[1, 2, 3, 4, 5]``` as the result (not necessarily in that order, taking in account the distributed nature of the record.).

### PR DESCRIPTION
**Pull request by priyank23**
This PR is for issue [#15388](https://github.com/chapel-lang/chapel/issues/15388).

Now, chapel has the following feature.
```use DistributedBag;

var bag = new DistBag(int);
bag.addBulk(1..5);
writeln(bag);
```
Output: ```[1, 3, 5, 2, 4]```
The order of the elements in the output may vary from one system to another.

Closes #15388

### CODE REVIEW

**Comment by priyank23** [comment]
@ben-albrecht Kindly review.

File: rst/users-guide/base/examples [hunk 1]
```diff
@@ -0,0 +1 @@
+/home/lp2917/Documents/gsoc_20/chapel/test/release/examples
```
Comment (ben-albrecht) [inline rst/users-guide/base/examples:1]: Please remove these files.

File: modules/packages/DistributedBag.chpl [hunk 2]
```diff
@@ -254,6 +254,12 @@ module DistributedBag {
       return chpl_getPrivatizedCopy(unmanaged DistributedBagImpl(eltType), _pid);
     }
 
+    proc readWriteThis(ch: channel) throws {
```
Comment (ben-albrecht) [inline modules/packages/DistributedBag.chpl:257]: Please add a test for this behavior (https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/ContributorInfo.rst#add-new-tests)

Comment (priyank23) [reply modules/packages/DistributedBag.chpl:257]: @ben-albrecht Can you point me to the directory where the tests for DistributedBags are done? I cannot find it at ```test/library/packages```

Comment (ben-albrecht) [reply modules/packages/DistributedBag.chpl:257]: Sure, they are in `test/library/packages/Collection`

**Comment by Aniket21mathur** [comment]
@priyank23 the `make` checks seem to fail for your last commit. Please see the details of why the build failed and fix them. Thanks :-)

**Comment by priyank23** [comment]
@Aniket21mathur That was an unusual error. It was showing "Found tabs :-(" but was working fine in my system. I thought chapel compiler automatically ignores tabs. Although, I have rendered that mistake, it is worth driving a discussion over this in the community. Any thoughts? @ben-albrecht

File: modules/packages/DistributedBag.chpl (hunk 2 above)
Comment (priyank23) [reply modules/packages/DistributedBag.chpl:259]: @ben-albrecht Kindly review.

**Comment by Aniket21mathur** [comment]
> @Aniket21mathur That was an unusual error. It was showing "Found tabs :-(" but was working fine in my system. I thought chapel compiler automatically ignores tabs. Although, I have rendered that mistake, it is worth driving a discussion over this in the community.

@priyank23 I think you are having a confusion between the chapel compiler and the travis CI build. I agree that the chapel compiler ignores tabs, but to have consistency the travis ci is configured to check for extra tabs. You can also run travis make tests locally by

`./util/buildRelease/smokeTest`

Please see `.travis.yml` present in the repo for more details.

Though it will be great to have a second opinion on this.

Thanks :-)

**Comment by Aniket21mathur** [comment]
@priyank23 I also reviewed the tests you have added, handle border cases and seems to look fine to me :-)

**Comment by priyank23** [comment]
> @priyank23 I think you are having a confusion between the chapel compiler and the travis CI build. I agree that the chapel compiler ignores tabs, but to have consistency the travis ci is configured to check for extra tabs.

@Aniket21mathur I did not know that. Thanks for stating that. :)

>@priyank23 I also reviewed the tests you have added, handle border cases and seems to look fine to me :-)

There will be only two cases here - no elements and some elements. The code and the tests support both of them.

File: modules/packages/DistributedBag.chpl [hunk 3]
```diff
@@ -253,7 +254,20 @@ module DistributedBag {
       }
       return chpl_getPrivatizedCopy(unmanaged DistributedBagImpl(eltType), _pid);
     }
-
+  
+    // printing the contents of DistBag onto a channel
+    proc readWriteThis(ch: channel) throws {
+      ch <~> "[";
+      var iterations=0;
```
Comment (ben-albrecht) [inline modules/packages/DistributedBag.chpl:261]: nit: spacing

```suggestion
      var iterations = 0;
```

File: modules/packages/DistributedBag.chpl [hunk 4]
```diff
@@ -253,7 +254,20 @@ module DistributedBag {
       }
       return chpl_getPrivatizedCopy(unmanaged DistributedBagImpl(eltType), _pid);
     }
-
+  
+    // printing the contents of DistBag onto a channel
```
Comment (ben-albrecht) [inline modules/packages/DistributedBag.chpl:258]: Let's make this a docstring:


```suggestion
    /* Read/write the contents of DistBag from/to a channel */
```

but also lets make sure this method doesn't show up in the documentation

```chpl
   pragma "no doc"
    /* Read/write the contents of DistBag from/to a channel */
   proc readWriteThis(ch: channel) throws {
```

File: modules/packages/DistributedBag.chpl [hunk 5]
```diff
@@ -253,7 +254,20 @@ module DistributedBag {
...
+    proc readWriteThis(ch: channel) throws {
+      ch <~> "[";
+      var iterations=0;
+      var size = this.getSize();
+      for i in this {
```
Comment (ben-albrecht) [inline modules/packages/DistributedBag.chpl:263]: You could alternatively do:

```suggestion
      for (i, iteration) in zip(this, 0..<this.getSize()) { {
```

Then you wouldn't have to create and update `iterations` in the loop

**Comment by ben-albrecht** [comment]
@priyank23 - the test added in this PR resulted in a failure due to distributed bags being unordered and `bag.add(1..2);` being promoted to adding `1` and `2` and parallel.

I think we will want to split the `bag.add(1..2)` into `bag.add(1);` and `bag.add(2);`. Are you able to make this change? If I don't hear back from you in a few hours, I'll make the change (to ensure clean nightly testing tonight).

Thanks!

**Comment by ben-albrecht** [comment]
Fixed here: https://github.com/chapel-lang/chapel/pull/15536

**Comment by priyank23** [comment]
@ben-albrecht My apologies for not being active when required. Thanks for making the change, though.

**Pull request merged**